core/domain/topic_domain.py

```diff
@@ -708,6 +713,35 @@ def delete_canonical_story(self, story_id):
                 'The story_id %s is not present in the canonical '
                 'story references list of the topic.' % story_id)
 
+    def rearrange_canonical_story(self, from_index, to_index):
+        """Rearranges or moves a canonical story to another position and
+
+        Args:
+            from_index: int. The index of canonical story index to move.
+            to_index: int. The index at which to insert the moved canonical
+                story.
+
+        Raises:
+            Exception. Invalid input.
+        """
+        if not isinstance(from_index, int):
+            raise Exception('Expected from_index value to be a number, '
+                            'received %s' % from_index)
+
+        if not isinstance(to_index, int):
+            raise Exception('Expected to_index value to be a number, '
+                            'received %s' % to_index)
+
+        if from_index == to_index:
+            raise Exception('Expected from_index and to_index values '
+                            'to be different.')
+
+        canonical_story_reference_to_move = copy.deepcopy(
```

Comment:
What if from_index and to_index are out of scope (< 0 or >= length)? Check that as well.
Add tests for those also.

Reply:
Done